### PR TITLE
[SUBGRAPH] Subgraph Matchstick Unit Tests

### PR DESCRIPTION
**Pull request by 0xdavinchee**
tl;dr
- Mapping logic changes are in `src/mappings/*.ts` and `src/*.ts` files
- This PR is a checkpoint so it's more reviewable
- Using Matchstick to run unit tests for mapping logic
- A good chunk of event entities implemented + validated
- Some HOL and aggregate entities

## Overview

This is just an initial PR for the Subgraph Matchstick Unit Tests, you can check [here](https://github.com/superfluid-finance/protocol-monorepo/issues/1106) to see what has been covered so far. I wanted to submit this PR as a checkpoint.

## In this PR
- Structure of tests is set
- Testing a good % of event entities 
- Some HOL and aggregate entities tested
- Lots of helper functions
- Tests added to test subgraph call handler yml
- README updated
- `schema.graphql` subgraph documentation updated
- Refactor of `let` to `const`

### CODE REVIEW

File: packages/subgraph/src/mappingHelpers.ts [hunk 1]
```diff
@@ -101,6 +103,8 @@ export function getOrInitSuperToken(
         token.createdAtTimestamp = currentTimestamp;
         token.createdAtBlockNumber = block.number;
         token.isSuperToken = true;
+        token.name = "";
```
Comment (0xdavinchee) [inline packages/subgraph/src/mappingHelpers.ts:106]: logic change alert: set token name and symbol to empty string here

File: packages/subgraph/src/mappingHelpers.ts [hunk 2]
```diff
@@ -116,21 +120,26 @@ export function getOrInitSuperToken(
...
         let tokenStatistic = getOrInitTokenStatistic(tokenAddress, block);
-        tokenStatistic = updateTotalSupplyForNativeSuperToken(
-            token,
+        tokenStatistic = getInitialTotalSupplyForSuperToken(
             tokenStatistic,
             tokenAddress
         );
         tokenStatistic.save();
 
+        // Per our TokenStatistic Invariant: whenever we create TokenStatistic, we must create TokenStatisticLog
```
Comment (0xdavinchee) [inline packages/subgraph/src/mappingHelpers.ts:129]: logic change alert: create token statistic entity

File: packages/subgraph/src/mappingHelpers.ts [hunk 3]
```diff
@@ -80,17 +80,19 @@ export function getOrInitAccount(
  * We also create token stats in here if it doesn't exist yet.
  */
 export function getOrInitSuperToken(
+    event: ethereum.Event,
```
Comment (0xdavinchee) [inline packages/subgraph/src/mappingHelpers.ts:83]: logic change alert: passing event and triggeredByEventName - both used in token statistic log creation

File: packages/subgraph/src/mappingHelpers.ts [hunk 4]
```diff
@@ -152,20 +161,25 @@ export function getOrInitSuperToken(
...
     token.createdAtTimestamp = block.timestamp;
+    token.decimals = 0;
+    token.name = "";
+    token.symbol = "";
     token.createdAtBlockNumber = block.number;
     token.isSuperToken = false;
+    token.isNativeAssetSuperToken = false;
     token.isListed = false;
-    token = getTokenInfoAndReturn(token as Token);
+    token = handleTokenRPCCalls(token, resolverAddress);
```
Comment (0xdavinchee) [inline packages/subgraph/src/mappingHelpers.ts:182]: logic change alert: handleTokenRPCCalls to also check if underlying token is listed

File: packages/subgraph/src/mappingHelpers.ts [hunk 5]
```diff
@@ -152,20 +161,25 @@ export function getOrInitSuperToken(
...
-    let tokenId = tokenAddress.toHex();
+    const tokenId = tokenAddress.toHex();
     let token = new Token(tokenId);
 
     if (tokenAddress.equals(ZERO_ADDRESS)) {
         return;
     }
 
     token.createdAtTimestamp = block.timestamp;
+    token.decimals = 0;
```
Comment (0xdavinchee) [inline packages/subgraph/src/mappingHelpers.ts:175]: logic change alert: default decimals, name, symbol

File: packages/subgraph/src/mappingHelpers.ts [hunk 6]
```diff
@@ -152,20 +161,25 @@ export function getOrInitSuperToken(
...
+    token.name = "";
+    token.symbol = "";
     token.createdAtBlockNumber = block.number;
     token.isSuperToken = false;
+    token.isNativeAssetSuperToken = false;
```
Comment (0xdavinchee) [inline packages/subgraph/src/mappingHelpers.ts:180]: logic change alert: isNativeAssetSuperToken set to false for underlying token

File: packages/subgraph/src/mappings/resolver.ts [hunk 7]
```diff
@@ -67,6 +67,8 @@ export function handleSet(event: Set): void {
             token.save();
         }
     }
+    resolverEntry.targetAddress = event.params.target;
```
Comment (0xdavinchee) [inline packages/subgraph/src/mappings/resolver.ts:79]: logic change alert: always set target address for resolverEntry when handling set (must do this after the logic in the if clause above it)

File: packages/subgraph/src/utils.ts [hunk 8]
```diff
@@ -78,13 +78,14 @@ export function handleTokenRPCCalls(
...
 ): Token {
-    token = getIsListedToken(token, resolverAddress);
-
     // we must handle the case when the native token hasn't been initialized
     // there is no name/symbol, but this may occur later
     if (token.name.length == 0 || token.symbol.length == 0) {
         token = getTokenInfoAndReturn(token);
     }
+    
+    // we do getIsListedToken after getTokenInfoAndReturn because it requires the token symbol
```
Comment (0xdavinchee) [inline packages/subgraph/src/utils.ts:87]: logic change alert: change order of rpc calls

File: packages/subgraph/src/utils.ts [hunk 9]
```diff
@@ -109,34 +110,34 @@ export function getIsListedToken(
...
 
-export function updateTotalSupplyForNativeSuperToken(
-    token: Token,
+/**
+ * Gets and sets the total supply for TokenStatistic of a SuperToken upon initial creation
+ * @param tokenStatistic 
+ * @param tokenAddress 
+ * @returns TokenStatistic
+ */
+export function getInitialTotalSupplyForSuperToken(
```
Comment (0xdavinchee) [inline packages/subgraph/src/utils.ts:131]: logic change alert: change function to be called for all super tokens (not just custom)

**Review by 0xdavinchee — COMMENTED** [review]
Some comments to track the mapping logic changes at a high level

**Review by hellwolf — APPROVED** [review]

File: packages/subgraph/src/mappingHelpers.ts [hunk 10]
```diff
@@ -80,17 +80,19 @@ export function getOrInitAccount(
  * We also create token stats in here if it doesn't exist yet.
  */
 export function getOrInitSuperToken(
+    event: ethereum.Event,
     tokenAddress: Address,
-    block: ethereum.Block
+    triggeredByEventName: string
```
Comment (kasparkallas) [inline packages/subgraph/src/mappingHelpers.ts:85]: What's the difference between this and `event.parameters.name`?

File: packages/subgraph/src/mappingHelpers.ts (hunk 1 above)
Comment (kasparkallas) [reply packages/subgraph/src/mappingHelpers.ts:106]: Otherwise they're null or something?

File: packages/subgraph/src/mappingHelpers.ts (hunk 2 above)
Comment (kasparkallas) [reply packages/subgraph/src/mappingHelpers.ts:129]: :spaghetti:

File: packages/subgraph/tests/assertionHelpers.ts [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    const id = createEventID(eventName, event);
+    const order = getOrder(event.block.number, event.logIndex);
+
+    assert.fieldEquals(entityType, id, "id", id);
+    assert.fieldEquals(entityType, id, "blockNumber", event.block.number.toString());
+    assert.fieldEquals(entityType, id, "logIndex", event.logIndex.toString());
+    assert.fieldEquals(entityType, id, "order", order.toString());
+    assert.fieldEquals(entityType, id, "timestamp", event.block.timestamp.toString());
+    assert.fieldEquals(entityType, id, "transactionHash", event.transaction.hash.toHex());
+    assert.fieldEquals(entityType, id, "gasPrice", event.transaction.gasPrice.toString());
```
Comment (kasparkallas) [inline packages/subgraph/tests/assertionHelpers.ts:28]: I don't see `"name"` asserted.

File: packages/subgraph/src/mappingHelpers.ts (hunk 6 above)
Comment (kasparkallas) [reply packages/subgraph/src/mappingHelpers.ts:180]: When I check existing Subgraphs then it is `false` there too. This explicitness is forced by Matchstick and it shouldn't have effect on production?

File: packages/subgraph/src/mappings/resolver.ts (hunk 7 above)
Comment (kasparkallas) [reply packages/subgraph/src/mappings/resolver.ts:79]: I don't understand why this has to be set here. Doesn't `getOrInitResolverEntry` already handle it?

File: packages/subgraph/src/utils.ts (hunk 9 above)
Comment (kasparkallas) [reply packages/subgraph/src/utils.ts:131]: I wonder why this changed?

File: packages/subgraph/src/mappingHelpers.ts (hunk 1 above)
Comment (0xdavinchee) [reply packages/subgraph/src/mappingHelpers.ts:106]: yup - the matchstick unit tests run into an issue where it tries to retrieve the values which are casted as not null (as they are defined as not null in our `schema.graphql`)

File: packages/subgraph/src/mappingHelpers.ts (hunk 6 above)
Comment (0xdavinchee) [reply packages/subgraph/src/mappingHelpers.ts:180]: yup, same thing as above where it expects not null, but is null. it has no effect on production as all underlying are false currently and if not, should be.

File: packages/subgraph/src/mappingHelpers.ts (hunk 10 above)
Comment (0xdavinchee) [reply packages/subgraph/src/mappingHelpers.ts:85]: Working on the refractor now, using nameof(event) where applicable.

File: packages/subgraph/src/mappingHelpers.ts (hunk 2 above)
Comment (0xdavinchee) [reply packages/subgraph/src/mappingHelpers.ts:129]: wat spaghetti

File: packages/subgraph/src/utils.ts (hunk 9 above)
Comment (0xdavinchee) [reply packages/subgraph/src/utils.ts:131]: it didn't make sense to only execute this function for native asset/custom super tokens

File: packages/subgraph/src/mappings/resolver.ts (hunk 7 above)
Comment (0xdavinchee) [reply packages/subgraph/src/mappings/resolver.ts:79]: no - we don't set it in `getOrInitResolverEntry` because we have logic slightly above in the if block which relies on `targetAddress` not being set yet (for example removing a token from the resolver), which is why we do it at the end of the call.

File: packages/subgraph/tests/assertionHelpers.ts (hunk 11 above)
Comment (0xdavinchee) [reply packages/subgraph/tests/assertionHelpers.ts:28]: true, this should be there

**Review by surewht — APPROVED** [review]

**Review by kasparkallas — APPROVED** [review]
